quiche/src/recovery/bytes_in_flight.rs

```diff
@@ -0,0 +1,144 @@
+use std::time::Duration;
+use std::time::Instant;
+
+#[derive(Default)]
+pub struct BytesInFlight {
+    // Current bytes in flight.
+    bytes_in_flight: usize,
+
+    // Instant at which bytes_in_flight transitioned from 0 to >0.
+    // Set if if bytes_in_flight is currently >0 which indicates that
+    // the bytes in flight interval is currently "open".
+    bytes_in_flight_interval_start: Option<Instant>,
+
+    // Duration of the current open interval.
+    open_interval_time: Duration,
+
+    // Sum of closed interval durations seen so far.
+    closed_interval_time: Duration,
+}
+
+impl BytesInFlight {
+    // Add to bytes in flight.  Record the start time when
+    // bytes_in_flight was 0 at the beginning of the function.
+    pub(crate) fn add(&mut self, delta: usize, now: Instant) {
+        self.bytes_in_flight += delta;
+        if delta > 0 {
+            if self.bytes_in_flight_interval_start.is_some() {
+                self.update_in_flight_times(now);
+            } else {
+                self.bytes_in_flight_interval_start = Some(now);
+            }
+        }
+    }
+
+    // Substract from bytes in flight.  If bytes_in_flight drops to 0,
+    // end the current bytes_in_flight >0 interval.
+    pub(crate) fn subtract(&mut self, delta: usize, now: Instant) {
+        self.bytes_in_flight -= delta;
+        self.update_in_flight_times(now);
+    }
+
+    // Same as subtract, but gracefully handles cases where the delta >
+    // bytes_in_flight.
+    pub(crate) fn saturating_subtract(&mut self, delta: usize, now: Instant) {
+        self.bytes_in_flight = self.bytes_in_flight.saturating_sub(delta);
+        self.update_in_flight_times(now);
+    }
+
+    pub(crate) fn get(&self) -> usize {
+        self.bytes_in_flight
+    }
+
+    // Total time during which bytes_in_flight was > 0.
+    pub(crate) fn get_time(&self) -> Duration {
```

Comment:
nit: since time might be conflated with a timestamp

```suggestion
    pub(crate) fn get_duration(&self) -> Duration {
```